src/publication_cache.rs

```diff
@@ -0,0 +1,191 @@
+///    
+///    .. code-block:: C
+///
+///       ze_publication_cache_options_t opts = ze_publication_cache_options_default();
+///       ze_owned_publication_cache_t pub_cache = ze_declare_publication_cache(z_loan(s), z_keyexpr(expr), &opts);
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn ze_declare_publication_cache(
+    session: z_session_t,
+    keyexpr: z_keyexpr_t,
+    options: Option<&ze_publication_cache_options_t>,
+) -> ze_owned_publication_cache_t {
+    match session.upgrade() {
+        Some(s) => {
+            let keyexpr = keyexpr.deref().as_ref().map(|s| s.clone().into_owned());
+            if let Some(key_expr) = keyexpr {
+                let mut p = s.declare_publication_cache(key_expr);
+                if let Some(options) = options {
+                    p = p.history(options.history.into());
+                    p = p.queryable_allowed_origin(options.queryable_origin.into());
+                    if options.resources_limit != 0 {
+                        p = p.resources_limit(options.resources_limit)
+                    }
+                    if options.queryable_prefix.deref().is_some() {
+                        let queryable_prefix = options
+                            .queryable_prefix
+                            .deref()
+                            .as_ref()
+                            .map(|s| s.clone().into_owned());
+                        if let Some(queryable_prefix) = queryable_prefix {
+                            p = p.queryable_prefix(queryable_prefix)
+                        }
+                    }
+                }
+                match p.res_sync() {
+                    Ok(publication_cache) => ze_owned_publication_cache_t::new(publication_cache),
+                    Err(e) => {
+                        log::error!("{}", e);
+                        ze_owned_publication_cache_t::null()
+                    }
+                }
+            } else {
+                log::error!("{}", UninitializedKeyExprError);
+                ze_owned_publication_cache_t::null()
+            }
+        }
+        None => ze_owned_publication_cache_t::null(),
+    }
+}
+
+/// Constructs a null safe-to-drop value of 'ze_owned_publication_cache_t' type
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn ze_publication_cache_null() -> ze_owned_publication_cache_t {
+    ze_owned_publication_cache_t::null()
+}
+
+/// Returns ``true`` if `pub_cache` is valid.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn ze_publication_cache_check(pub_cache: &ze_owned_publication_cache_t) -> bool {
+    pub_cache.as_ref().is_some()
+}
+
+/// Closes the given :c:type:`ze_owned_publication_cache_t`, droping it and invalidating it for double-drop safety.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn ze_close_publication_cache(pub_cache: &mut ze_owned_publication_cache_t) -> i8 {
```

Comment:
Use `ze_undeclare_publication_cache` or `ze_publication_cache_drop` instead to stick with the 2 main conventions we have.